File: netDxf/IO/DxfWriter.StoredCellStyleMap.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System.IO;
using System.Linq;
using netDxf.Collections;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private bool WriteStoredCellStyleMapPayload(DxfDatabaseObject item)
        {
            if (!(item is DxfStoredCellStyleMap content)) return false;
            foreach (DxfTag tag in content.Payload) this.WriteDatabaseTag(tag, false);
            return true;
        }
        private void PrepareStoredCellStyleMapClass(DxfClassCollection definitions)
        {
            const string name = "CELLSTYLEMAP";
            int count = this.doc.Objects.Items.Count(item => item.CodeName == name);
            bool typed = this.doc.Objects.Items.Any(item => item is DxfStoredCellStyleMap);
            if (definitions.Contains(name))
            {
                var definition = definitions[name];
                if (definition.CppClassName != "AcDbCellStyleMap" || definition.IsEntity)
                { if (typed) throw new InvalidDataException("CLASS conflicts with stored CELLSTYLEMAP."); return; }
                if (typed || count == 0) definition.InstanceCount = count;
            }
            else if (typed) definitions.Add(new DxfClass(name, "AcDbCellStyleMap", "ObjectDBX Classes") { ProxyFlags = 1152, IsEntity = false, InstanceCount = count });
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.StoredCellStyleMap.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfStoredCellStyleMap } from '../Objects/DxfStoredCellStyleMap.js';
import { DxfClass } from '../DxfClass.js';
import { InvalidDataException } from '../../runtime/Errors.js';
export function WriteStoredCellStyleMapPayload(chunk,version,item){
  if(!(item instanceof DxfStoredCellStyleMap))return false;
  for(const tag of item.Payload)chunk.Write(tag.Code,tag.Value);return true;
}
export function PrepareStoredCellStyleMapClass(document,definitions){
  const count=Array.from(document.Objects.Items).filter(i=>i.CodeName==='CELLSTYLEMAP').length,typed=Array.from(document.Objects.Items).some(i=>i instanceof DxfStoredCellStyleMap);
  if(definitions.Contains('CELLSTYLEMAP')){const definition=definitions.get_Item('CELLSTYLEMAP');
    if(definition.CppClassName!=='AcDbCellStyleMap'||definition.IsEntity){if(typed)throw new InvalidDataException('CLASS conflicts with stored CELLSTYLEMAP.');return;}
    if(typed||count===0)definition.InstanceCount=count;
  }else if(typed){const definition=new DxfClass('CELLSTYLEMAP','AcDbCellStyleMap','ObjectDBX Classes');definition.ProxyFlags=1152;definition.IsEntity=false;definition.InstanceCount=count;definitions.Add(definition);}
}
